opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -0,0 +1,1462 @@
+                    if (rv > rvSat * (1+epsilon) ) {
+                        reservoir_state.hydroCarbonState()[cell_idx] = HydroCarbonState::GasAndOil;
+                        so = epsilon;
+                        rv = rvSat;
+                        sg -= epsilon;
+                    }
+                    break;
+                }
+
+                default:
+                    OPM_THROW(std::logic_error, "Unknown primary variable enum value in cell " << cell_idx << ": " << hydroCarbonState);
+                }
+            }
+
+        }
+
+
+        /// Return true if output to cout is wanted.
+        bool terminalOutputEnabled() const
+        {
+            return terminal_output_;
+        }
+
+        template <class CollectiveCommunication>
+        double convergenceReduction(const CollectiveCommunication& comm,
+                                    const long int ncGlobal,
+                                    const int np,
+                                    const std::vector< std::vector< Scalar > >& B,
+                                    const std::vector< std::vector< Scalar > >& tempV,
+                                    const std::vector< std::vector< Scalar > >& R,
+                                    const std::vector< Scalar >& pv,
+                                    const std::vector< Scalar >& residual_well,
+                                    std::vector< Scalar >& R_sum,
+                                    std::vector< Scalar >& maxCoeff,
+                                    std::vector< Scalar >& B_avg,
+                                    std::vector< Scalar >& maxNormWell )
+        {
+            const int nw = residual_well.size() / np;
+            assert(nw * np == int(residual_well.size()));
+
+            // Do the global reductions
+            B_avg.resize(np);
+            maxCoeff.resize(np);
+            R_sum.resize(np);
+            maxNormWell.resize(np);
+
+            // computation
+            for ( int idx = 0; idx < np; ++idx )
+            {
+                B_avg[idx] = std::accumulate( B[ idx ].begin(), B[ idx ].end(), 0.0 ) / double(ncGlobal);
+                R_sum[idx] = std::accumulate( R[ idx ].begin(), R[ idx ].end(), 0.0 );
+                maxCoeff[idx] = *(std::max_element( tempV[ idx ].begin(), tempV[ idx ].end() ));
+
+                assert(np >= np);
+                if (idx < np) {
+                    maxNormWell[idx] = 0.0;
+                    for ( int w = 0; w < nw; ++w ) {
+                        maxNormWell[idx] = std::max(maxNormWell[idx], std::abs(residual_well[nw*idx + w]));
+                    }
+                }
+            }
+
+            // Compute total pore volume
+            double pvSum = std::accumulate(pv.begin(), pv.end(), 0.0);
+
+            if( comm.size() > 1 )
+            {
```

Comment:
Why didn't reuse or copy/paste existing code that is known to work. This code does not account for shared cells and will cause wrong results in parallel.